### PR TITLE
allow shadow memory offset per shadow ioeventfd

### PR DESCRIPTION
**Pull request by tmakatos**
Signed-off-by: Thanos Makatos <thanos.makatos@nutanix.com>

### CODE REVIEW

File: include/libvfio-user.h [hunk 1]
```diff
@@ -1076,11 +1076,12 @@ vfu_sg_is_mappable(vfu_ctx_t *vfu_ctx, dma_sg_t *sg);
  *  Requires a kernel with shadow ioeventfd support.
  *  Experimental, must be compiled with SHADOW_IOEVENTFD defined, otherwise
  *  must be -1.
+ * @shadow_offset: offset in shadow memory where value is written to.
  */
 int
 vfu_create_ioeventfd(vfu_ctx_t *vfu_ctx, uint32_t region_idx, int fd,
-                     size_t offset, uint32_t size, uint32_t flags,
-                     uint64_t datamatch, int shadow_fd);
+                     size_t gpa_offset, uint32_t size, uint32_t flags,
```
Comment (jlevon) [inline include/libvfio-user.h:1083]: need to update :1068

File: lib/libvfio-user.c [hunk 2]
```diff
@@ -140,36 +140,31 @@ dev_get_caps(vfu_ctx_t *vfu_ctx, vfu_reg_info_t *vfu_reg, bool is_migr_reg,
...
-
         *fds = malloc(nr_mmap_areas * sizeof(int));
         if (*fds == NULL) {
             return ERROR_INT(ENOMEM);
         }
         sparse->header.id = VFIO_REGION_INFO_CAP_SPARSE_MMAP;
         sparse->header.version = 1;
         sparse->header.next = 0;
-        sparse->nr_areas = *nr_fds = nr_mmap_areas;
+        sparse->nr_areas = nr_mmap_areas;
```
Comment (jlevon) [inline lib/libvfio-user.c:150]: split out this fix into a separate PR?

Comment (jlevon) [reply lib/libvfio-user.c:150]: also could add a test maybe?

File: lib/libvfio-user.c [hunk 3]
```diff
@@ -236,13 +232,15 @@ region_access(vfu_ctx_t *vfu_ctx, size_t region, char *buf,
     }
 
 out:
-    if (ret != (ssize_t)count) {
+    if (unlikely(ret != (ssize_t)count)) {
         vfu_log(vfu_ctx, LOG_DEBUG, "region%zu: %s (%#llx:%zu) failed: %m",
                 region, verb, (ull_t)offset, count);
-    } else {
+    }
+#ifdef DEBUG
```
Comment (jlevon) [inline lib/libvfio-user.c:239]: instead can you define:

#else
static void debug_region_access(..)
{}
#endif

when non-debug?

File: lib/libvfio-user.c [hunk 4]
```diff
@@ -255,31 +253,33 @@ is_valid_region_access(vfu_ctx_t *vfu_ctx, size_t size, uint16_t cmd,
...
 
-    if (ra->count > SERVER_MAX_DATA_XFER_SIZE) {
+    if (unlikely(ra->count > SERVER_MAX_DATA_XFER_SIZE)) {
         vfu_log(vfu_ctx, LOG_ERR, "region access count too large (%u)",
                 ra->count);
         return false;
     }
 
-    if (cmd == VFIO_USER_REGION_WRITE && size - sizeof(*ra) != ra->count) {
+    if (unlikely(cmd == VFIO_USER_REGION_WRITE
```
Comment (jlevon) [inline lib/libvfio-user.c:267]: nit, && on first line please

File: lib/libvfio-user.c [hunk 5]
```diff
@@ -311,11 +311,11 @@ handle_region_access(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg)
     assert(vfu_ctx != NULL);
     assert(msg != NULL);
 
-    if (!is_valid_region_access(vfu_ctx, msg->in.iov.iov_len, msg->hdr.cmd, in_ra)) {
+    if (unlikely(!is_valid_region_access(vfu_ctx, msg->in.iov.iov_len, msg->hdr.cmd, in_ra))) {
```
Comment (jlevon) [inline lib/libvfio-user.c:314]: line length?

File: lib/libvfio-user.c (hunk 2 above)
Comment (tmakatos) [reply lib/libvfio-user.c:150]: > split out this fix into a separate PR?

yes

> also could add a test maybe?

We already have `test_device_get_region_info_caps()`, I'll extend it.

File: include/libvfio-user.h (hunk 1 above)
Comment (tmakatos) [reply include/libvfio-user.h:1083]: done

File: lib/libvfio-user.c (hunk 5 above)
Comment (tmakatos) [reply lib/libvfio-user.c:314]: ok

File: lib/libvfio-user.c (hunk 4 above)
Comment (tmakatos) [reply lib/libvfio-user.c:267]: ok

File: lib/libvfio-user.c (hunk 3 above)
Comment (tmakatos) [reply lib/libvfio-user.c:239]: done

**Comment by tmakatos** [comment]
I should also move the `debug_region_access()` and unlikely fixes into separate commits.

**Comment by tmakatos** [comment]
@jlevon this is ready to be reviewed, it only contains code for specific the offset in the shadow memory.

**Review by jlevon — APPROVED** [review]
